### PR TITLE
Add Tab component

### LINKED ISSUE #143: [Change Request]: Add Tab component
### What type of enhancement/feature is this?

New UI Component, Documentation Improvements

### Description

Add Tab component and possibly Tab Group component to Omni-component library

- Add Tab component 
- Add Tab group component
- Add various properties and states
- Add the required documentation for various templates
- Add play functions where required

### Use Case/Reasoning for enhancement/feature.

The tab component is a required component for most UIs which will allow a user to tab between grouped content to render. 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have [searched for existing issues](https://github.com/capitec/omni-components/issues) and to the best of my knowledge this is not a duplicate
- [X] This is a single enhancement/feature

### PR DESCRIPTION
**Pull request by chromaticWaster**
### Description:

closes #143 

- Added TabGroup, Tab and Tab Header component to Omni-components.
- Added story examples with play functions.
- Tested on Desktop browsers and Mobile devices.

(*Link to bug/feature from [Issues](../issues) and how you resolve it*)

### Screenshots (if appropriate):

image

### All Submissions:

* [x] I have followed the [contribution guidelines](https://github.com/capitec/omni-components/blob/develop/CONTRIBUTING.md) for this project.
* [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.

### Changes to Core Features:

* [x] I have added an explanation of what my changes do and why I would like to include them.
* [x] I have written new tests for my core changes, as applicable.
* [x] I have successfully ran tests with my changes locally.
* [x] I have added/updated docs, as applicable.

### CODE REVIEW

File: src/tab/Tabs.stories.ts
Comment (BOTLANNER) [inline src/tab/Tabs.stories.ts:1]: Please consider what actually is being tested with the play functions, they seem to be doing very generic checks but not actually testing for something relating to the story

File: src/tab/Tabs.ts [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+ *   <omni-label>Tab 3</omni-label>
+ *  </omni-tab>
+ * </omni-tabs>
+ *```
+ *
+ * @element omni-tabs
+ *
+ * Registry of all properties defined by the component.
+ *
+ * @fires {CustomEvent<{}>} tab-select - Dispatched when one of the omni-tab(s) is selected.
```
Comment (BOTLANNER) [inline src/tab/Tabs.ts:34]:
```suggestion
 * @fires {CustomEvent<{ previous: HTMLElement, selected: HTMLElement}>} tab-select - Dispatched when an omni-tab is selected.
```

File: src/tab/Tabs.ts [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ *
+ * @slot - Content to render inside the component body.
```
Comment (BOTLANNER) [inline src/tab/Tabs.ts:36]:
```suggestion
 * @slot - All omni-tab components that are managed by this component.
```

File: src/tab/Tabs.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+ * @slot header - The header slot used to slot omni-tab-header components.
```
Comment (BOTLANNER) [inline src/tab/Tabs.ts:37]:
```suggestion
 * @slot header - Optional omni-tab-header components associated with each omni-tab component.
```

File: src/tab/Tabs.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+ * @cssprop --omni-tab-header-indicator-border-radius - Tab header component indicator border radius.
+ * @cssprop --omni-tab-header-indicator-width - Tab header component indicator width.
+ */
+@customElement('omni-tabs')
+export class Tabs extends OmniElement {
+    @state() private _observer: MutationObserver | undefined;
+
+    override connectedCallback(): void {
+        super.connectedCallback();
+        this._observer = new MutationObserver((mutations) => {
```
Comment (BOTLANNER) [inline src/tab/Tabs.ts:74]: What is the browser support for `MutationObserver`?
Should this not be wrapped with `if("MutationObserver" in window)` to cater for older browsers?

File: src/tab/Tabs.ts [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    selectTab(tabHeader: TabHeader) {
+        if (!tabHeader || tabHeader.classList.contains('tab-bar')) {
+            return;
+        }
+
+        // set tab header used in cases where the tab header has slotted content.
+        tabHeader = tabHeader.closest('omni-tab-header') as TabHeader;
+        const children = Array.from(this.children);
+
+        const tab = children.find((t) => (t.id === tabHeader.for && t.id !== '' && tabHeader.for !== '') || t === tabHeader.data);
```
Comment (BOTLANNER) [inline src/tab/Tabs.ts:117]:
```suggestion
        const tab = children.find((t) => (t.id && t.id === tabHeader.for) || t === tabHeader.data);
```

**Review by BOTLANNER — CHANGES_REQUESTED** [review]

File: src/tab/Tab.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
+import { css, html, TemplateResult } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { OmniElement } from '../core/OmniElement.js';
+
+/**
+ * Control that can be used to display slotted content used within a omni-tabs component.
```
[the remark this answers is not in the document]
Comment (chromaticWaster) [reply src/tab/Tab.ts:6]: Updated as per the suggestion

File: src/tab/TabHeader.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
+import { css, html, nothing, TemplateResult } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { OmniElement } from '../core/OmniElement.js';
+
+/**
+ * Control that can be used to display custom slotted content used within a omni-tabs component with an associated omni-tab component.
```
[the remark this answers is not in the document]
Comment (chromaticWaster) [reply src/tab/TabHeader.ts:6]: Updated as per the suggestion

File: src/tab/TabHeader.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+ *
+ * @import
+ * ```js
+ * import '@capitec/omni-components/tab';
+ * ```
+ * @example
+ * html```
+ * <omni-tab-header for="Tab 1">
```
[the remark this answers is not in the document]
Comment (chromaticWaster) [reply src/tab/TabHeader.ts:14]: Updated as per suggestion

File: src/tab/Tabs.stories.ts [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+        await expect(tabBar).toBeTruthy();
+
+        // Get all the tabs in the tab bar
+        const nestedTabs = tabBar.querySelectorAll('omni-tab-header');
+        await expect(nestedTabs).toBeTruthy();
+        const tabsArray = [...nestedTabs];
+
+        //Get the active tab.
+        const activeTab = tabsArray.find((c) => c.hasAttribute('data-active'));
+        await expect(activeTab).toBeTruthy;
```
[the remark this answers is not in the document]
Comment (chromaticWaster) [reply src/tab/Tabs.stories.ts:83]: Updated

File: src/tab/Tabs.ts (hunk 1 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts:34]: Updated as per the suggestion

File: src/tab/Tabs.ts (hunk 2 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts:36]: Updated as per suggestion

File: src/tab/Tabs.ts (hunk 3 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts:37]: Updated as per suggestion

File: src/tab/Tabs.ts (hunk 4 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts:74]: Browser support for most browsers dates back to 2013.

File: src/tab/Tabs.ts (hunk 5 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts]: Updated as per suggestion

File: src/tab/Tabs.stories.ts [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+<omni-tab header='Tab 2'>
+    <div>Tab 2 Content</div>
+</omni-tab>
+<omni-tab header='Tab 3'>
+    <div>Tab 3 Content</div>
+</omni-tab>`
+    },
+    play: async (context) => {
+        const tabs = within(context.canvasElement).getByTestId<Tabs>('test-tabs');
+        const click = jest.fn();
```
[the remark this answers is not in the document]
Comment (chromaticWaster) [reply src/tab/Tabs.stories.ts]: Removed

File: src/tab/Tabs.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+ *
+ * @slot - All omni-tab components that are managed by this component.
+ * @slot header - Optional omni-tab-header components associated with each omni-tab component.
+ *
+ * @cssprop --omni-tabs-tab-bar-width - Tabs tab bar width.
+ * @cssprop --omni-tabs-tab-bar-height - Tabs tab bar height.
+ * @cssprop --omni-tabs-tab-bar-border-bottom - Tabs tab bar bottom border.
+ * @cssprop --omni-tabs-tab-bar-background-color - Tabs tab bar background color.
+ *
+ * @cssprop --omni-tab-header-font-color - Tab header component font color.
```
Comment (stefan505) [inline src/tab/Tabs.ts:44]: Are all header related props correct to list here?

File: src/tab/Tab.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    @property({ type: String, reflect: true }) header?: string;
+
+    /**
+     * Indicator if the component is disabled.
+     * @attr
+     */
+    @property({ type: Boolean, reflect: true }) disabled?: boolean;
+
+    /**
+     * Indicator if the component is active.
```
Comment (stefan505) [inline src/tab/Tab.ts:39]:
```suggestion
     * Indicator if the tab is active.
```

File: src/tab/Tab.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+@customElement('omni-tab')
+export class Tab extends OmniElement {
+    /**
+     * Tab header label.
+     * @attr
+     */
+    @property({ type: String, reflect: true }) header?: string;
+
+    /**
+     * Indicator if the component is disabled.
```
Comment (stefan505) [inline src/tab/Tab.ts:33]:
```suggestion
     * Indicator if the tab is disabled.
```

File: src/tab/Tab.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @element omni-tab
+ *
+ * @slot - Content to render inside the component body.
+ *
+ */
+@customElement('omni-tab')
+export class Tab extends OmniElement {
+    /**
+     * Tab header label.
```
Comment (stefan505) [inline src/tab/Tab.ts:27]: Maybe mention here the option to use the header component as well for more complex header layouts.

File: src/tab/Tab.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+     * Indicator if the component is active.
+     * @attr
+     */
+    @property({ type: Boolean, reflect: true }) active?: boolean;
+
+    static override get styles() {
+        return [
+            super.styles,
+            css`
+            :host {
```
Comment (stefan505) [inline src/tab/Tab.ts:48]: Check code indentation here. Lines 48 to 56 needs to be indented one tab.

File: src/tab/TabHeader.ts [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+     * Indicator of which omni-tab element with the matching corresponding id attribute should be displayed.
+     * @attr
+     */
+    @property({ type: String, reflect: true }) for?: string;
+
+    /**
+     * Data associated with the component.
+     * @no_attribute
+     */
+    @property({ type: Object, reflect: false }) data?: unknown;
```
Comment (stefan505) [inline src/tab/TabHeader.ts:59]: This needed?

File: src/tab/TabHeader.ts [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+    static override get styles() {
+        return [
+            super.styles,
+            css`
```
Comment (stefan505) [inline src/tab/TabHeader.ts:64]: Check code indentation inside `css` template result.

File: src/tab/Tabs.ts [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+        );
+
+        this.requestUpdate();
+    }
+
+    static override get styles() {
+        return [
+            super.styles,
+            css`
+            :host {
```
Comment (stefan505) [inline src/tab/Tabs.ts:160]: Check code indentation inside the `css` template function.

File: src/tab/Tabs.stories.ts [hunk 19]
```diff
@@ -0,0 +1,445 @@
...
+                )
+        }
+    ],
+    name: 'Interactive',
+    description: () => html`
+        <p>
+        The <code class="language-html">&lt;omni-tabs&gt;</code> component will display content based on the nested <code class="language-html">&lt;omni-tab&gt;</code> component(s). 
+        </p>
+        <p>
+        Headers are managed by either setting the <code>header</code> attribute of the <code class="language-html">&lt;omni-tab&gt;</code> component or via slotted <code class="language-html">&lt;omni-tab-header&gt;</code> component(s) that targets the <code class="language-html">&lt;omni-tabs&gt;</code> header slot by setting the <code>slot</code> attribute to <code class="language-js">"header"</code>, and displays the slotted content of the <code class="language-html">&lt;omni-tab&gt;</code> with the <code>active</code> attribute set.
```
Comment (stefan505) [inline src/tab/Tabs.stories.ts:55]: This is very technical, wall of text. We need to find a way to convey what it does, instead of how it does it.

**Review by stefan505 — CHANGES_REQUESTED** [review]

File: src/tab/Tabs.ts (hunk 11 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts:44]: listed here to make these css variables visible on the doc site as there is not a stories file for the tab header

File: src/tab/Tabs.ts (hunk 18 above)
Comment (chromaticWaster) [reply src/tab/Tabs.ts]: Updated

File: src/tab/TabHeader.ts (hunk 17 above)
Comment (chromaticWaster) [reply src/tab/TabHeader.ts:64]: Updated

File: src/tab/TabHeader.ts (hunk 16 above)
Comment (chromaticWaster) [reply src/tab/TabHeader.ts:59]: Yes it is used to associate a tab with a specific tab header.

File: src/tab/Tab.ts (hunk 15 above)
Comment (chromaticWaster) [reply src/tab/Tab.ts]: Updated

File: src/tab/Tab.ts (hunk 13 above)
Comment (chromaticWaster) [reply src/tab/Tab.ts]: Updated

File: src/tab/Tab.ts (hunk 12 above)
Comment (chromaticWaster) [reply src/tab/Tab.ts]: Updated

File: src/tab/Tab.ts (hunk 14 above)
Comment (chromaticWaster) [reply src/tab/Tab.ts]: Updated

File: src/tab/Tabs.stories.ts (hunk 19 above)
Comment (chromaticWaster) [reply src/tab/Tabs.stories.ts]: Will remove the statement regarding the active tab as that can be conveyed by the active story. I will update the description accordingly feel free to suggest amendments.

File: src/tab/Tabs.ts (hunk 11 above)
Comment (stefan505) [reply src/tab/Tabs.ts:44]: That's not great, we should extend the docs site to cope with the implementation, rather hack the component to work on the docs site. I did try to bring this up initially when we still had the original components.

Comment (chromaticWaster) [reply src/tab/Tabs.ts]: Added a change to the StoryUtils.ts file to render the css variables of the tab header and tab on the tabs docs.

File: src/tab/Tabs.stories.ts [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+    name: 'Interactive',
+    description: () => html`
+        <p>
+        The <code class="language-html">&lt;omni-tabs&gt;</code> component will display content based on the nested <code class="language-html">&lt;omni-tab&gt;</code> component(s). 
+        </p>
+        <p>
+        Tab headers are rendered by either setting the <code>header</code> attribute of the <code class="language-html">&lt;omni-tab&gt;</code> component or via slotted <code class="language-html">&lt;omni-tab-header&gt;</code> component(s) that targets the <code class="language-html">&lt;omni-tabs&gt;</code> header slot. 
+        </p>
+    `,
+    args: {
```
Comment (stefan505) [inline src/tab/Tabs.stories.ts:58]: The string contained in `args` for the default slot has single quotes for attribute values. Should be double quotes for consistency.

Comment (chromaticWaster) [reply src/tab/Tabs.stories.ts:47]: Updated as part of the Tab components story file split

**Review by BOTLANNER — APPROVED** [review]

File: src/tab/TabGroup.ts [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+ * @slot - All omni-tab components that are managed by this component.
+ * @slot header - Optional omni-tab-header components associated with each omni-tab component.
+ *
+ * @cssprop --omni-tabs-tab-bar-width - Tabs tab bar width.
+ * @cssprop --omni-tabs-tab-bar-height - Tabs tab bar height.
+ * @cssprop --omni-tabs-tab-bar-border-bottom - Tabs tab bar bottom border.
+ * @cssprop --omni-tabs-tab-bar-background-color - Tabs tab bar background color.
+ *
+ * @cssinherit omni-tab-header
+ * @cssinherit omni-tab
```
Comment (stefan505) [inline src/tab/TabGroup.ts:45]:
```suggestion
```

File: src/tab/TabGroup.ts [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+ * @element omni-tab-group
+ *
+ * Registry of all properties defined by the component.
+ *
+ * @fires {CustomEvent<{ previous: HTMLElement, selected: HTMLElement}>} tab-select - Dispatched when an omni-tab is selected.
+ *
+ * @slot - All omni-tab components that are managed by this component.
+ * @slot header - Optional omni-tab-header components associated with each omni-tab component.
+ *
+ * @cssprop --omni-tabs-tab-bar-width - Tabs tab bar width.
```
Comment (stefan505) [inline src/tab/TabGroup.ts:39]:
```suggestion
 * @cssprop --omni-tab-group-tab-bar-width - Tabs tab bar width.
```

Appy, to all "-tabs" related names.

File: src/tab/Tab.ts [hunk 23]
```diff
@@ -0,0 +1,72 @@
+import { css, html, TemplateResult } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { OmniElement } from '../core/OmniElement.js';
+
+/**
+ * Control that can be used to display slotted content, for use within an omni-tab-group component.
```
Comment (stefan505) [inline src/tab/Tab.ts:6]:
```suggestion
 * Control that can be used to display slotted content, for use within an Tab Group component.
```

File: src/tab/Tab.stories.ts [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+ <OmniTab header="Tab 3">
+    <OmniLabel label="Label of Tab 3"/>
+ </OmniTab>
+</OmniTabGroup>;`
+        }
+    ],
+    name: 'Basic',
+    description: () => html`
+    <div>
+        This is the recommended use of the <code class="language-html">&lt;omni-tab-group&gt;</code> with slotted <code class="language-html">&lt;omni-tab&gt;</code> component(s), headers for each tab are set by setting the <code>header</code> attribute of the <code class="language-html">&lt;omni-tab&gt;</code>.
```
Comment (stefan505) [inline src/tab/Tab.stories.ts:58]:
```suggestion
        This is the recommended use. Headers for each tab is set by the <code>header</code> attribute.
```

File: src/tab/Tab.stories.ts [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+     <OmniLabel label='Label of Tab 3'/>
+ </OmniTab>
+</OmniTabGroup>;`
+        }
+    ],
+    args: {},
+    name: 'Active',
+    description: () => html`
+    <div>
+        Set which <code class="language-html">&lt;omni-tab&gt;</code> slotted in the <code class="language-html">&lt;omni-tab-group&gt;</code> should be active by default. By setting the <code>active</code> attribute of one of the slotted <code class="language-html">&lt;omni-tab&gt;</code> components.
```
Comment (stefan505) [inline src/tab/Tab.stories.ts:126]:
```suggestion
        Set the <code>active</code> attribute on an <code class="language-html">&lt;omni-tab&gt;</code> to indicate its active. By default, the first slotted one is active.
```

File: src/tab/Tab.stories.ts [hunk 26]
```diff
@@ -0,0 +1,232 @@
...
+ <OmniTab header="Tab 3" disabled>
+    <OmniLabel label="Label of Tab 3"/>
+ </OmniTab>
+</OmniTabGroup>;`
+        }
+    ],
+    name: 'Disabled',
+    description: () => html`
+    <div>
+        Set a <code class="language-html">&lt;omni-tab&gt;</code> slotted in the <code class="language-html">&lt;omni-tab-group&gt;</code> component to be disabled by setting <code>disabled</code> attribute.
```
Comment (stefan505) [inline src/tab/Tab.stories.ts:201]:
```suggestion
        Set the <code>disabled</code> attribute on an <code class="language-html">&lt;omni-tab&gt;</code> to indicate its disabled.
```

File: src/tab/TabGroup.stories.ts [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            framework: 'Vue',
+            load: (args) =>
+                getSourceFromLit(Interactive!.render!(args), undefined, (s) =>
+                    s.replace(' active', ' :active="true"').replace(' disabled', ' :disabled="true"')
+                )
+        }
+    ],
+    name: 'Interactive',
+    description: () => html`
```
Comment (stefan505) [inline src/tab/TabGroup.stories.ts:39]: Simplify 2nd sentence to link to Tab Header for advanced use case of slotting.

File: src/tab/TabHeader.stories.ts [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+        </omni-tab>
+        <omni-tab id="left">
+            <div>Left</div>
+        </omni-tab>
+        <omni-tab id="right">
+            <div>Right</div>
+        </omni-tab>
+        `
+    },
+    description: () => html`
```
Comment (stefan505) [inline src/tab/TabHeader.stories.ts:105]: Wrap "header" with code, e.g. `slot="header"`

Comment (stefan505) [inline src/tab/TabHeader.stories.ts:105]: Wrap "header" with code, e.g. `slot="header"`

File: src/tab/TabHeader.ts [hunk 29]
```diff
@@ -0,0 +1,146 @@
+import { css, html, nothing, TemplateResult } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { OmniElement } from '../core/OmniElement.js';
+
+/**
+ * Control that can be used to display custom slotted content, for use within an omni-tabs component and associated omni-tab component.
```
Comment (stefan505) [inline src/tab/TabHeader.ts:6]:
```suggestion
 * Control that can be used to display custom slotted content, for use within Tab Group component with associated Tab component.
```

**Review by stefan505 — CHANGES_REQUESTED** [review]

File: src/tab/TabHeader.ts (hunk 29 above)
Comment (chromaticWaster) [reply src/tab/TabHeader.ts:6]: Updated

File: src/tab/Tab.stories.ts (hunk 26 above)
Comment (chromaticWaster) [reply src/tab/Tab.stories.ts:201]: Updated

File: src/tab/Tab.stories.ts (hunk 25 above)
Comment (chromaticWaster) [reply src/tab/Tab.stories.ts:126]: Updated

File: src/tab/Tab.stories.ts (hunk 24 above)
Comment (chromaticWaster) [reply src/tab/Tab.stories.ts:58]: Updated

File: src/tab/TabGroup.ts (hunk 22 above)
Comment (chromaticWaster) [reply src/tab/TabGroup.ts:39]: Updated

File: src/tab/TabGroup.ts (hunk 21 above)
Comment (chromaticWaster) [reply src/tab/TabGroup.ts:45]: Updated

File: src/tab/TabHeader.stories.ts (hunk 28 above)
Comment (chromaticWaster) [reply src/tab/TabHeader.stories.ts:105]: Updated

Comment (chromaticWaster) [reply src/tab/TabHeader.stories.ts:105]: Updated

**Pull request merged**